add transcript events: init, snapshot save/load, shutdown

warner opened the pull request:
This introduces four new pseudo-delivery events to the transcript:

* 'initialize-worker': a new empty worker is created
* 'load-snapshot': a worker is loaded from heap snapshot
* 'save-snapshot': we tell the worker to write a heap snapshot
* 'shutdown-worker': we stop the worker (e.g. during upgrade)

These events are not actually delivered to the worker: they are not VatDeliveryObjects. However many of them are implemented with commands to the worker (just not `deliver()` commands). The vat-warehouse records these events in the transcript to help
subsequent (manual/external) replay tools know what happened. Without them, we'd need to deduce e.g. the heap-snapshot writing schedule by counting deliveries and comparing them against
snapshotInitial/snapshotInterval .

The 'save-snapshot'/'load-snapshot' pair indicates what a replay would do. It does not mean that the vat-warehouse actually tore down the old worker and immediately replaced it with a new one (from snapshot). It might choose to do that, or the worker itself might choose to replace its XS engine instance with a fresh one, or it might keep using the old engine. The 'save-snapshot' command has side-effects (it does a forced GC), so it is important to keep track of when it happened.

The transcript is broken up into "spans", delimited by heap snapshots or upgrade-related shutdowns. To bring a worker up to date, we want to start a worker (either a blank one, or from a snapshot), and then replay the "current span".

With this change, the current span always starts either with 'initialize-worker' or with 'load-snapshot', telling us exactly what needs to be done. The span then contains all the deliveries that must be replayed. The current span will never include a 'save-snapshot' or 'shutdown-worker': the span is closed immediately after those events are added, so replay will never see them. But a tool which replays a historical span will see them at the end.

The types were improved to make `TranscriptDelivery` be a superset of `VatDeliveryObject`. We also record TranscriptDeliveryResult, which is currently a stripped down subset of VatDeliveryResult (just the "ok" status), except that save-snapshot includes the snapshot hash in its results. In the future, we'll probably record the deterministic subset of metering results (computrons, maybe something about memory allocation).

refs #7199
refs #6770

warner commented:
w.r.t. #6770:
* it records the schedule of heap-snapshot saves
* it does not record the computrons included in a delivery

I think we're in agreement that computrons should be in-consensus, so adding a `TranscriptDeliveryGenericResults.metering.computrons` would be appropriate.

w.r.t. #7199
* it adds a transcript entry when initializing the worker
* it does not record the helper bundles (lockdown+supervisor) used to start a new worker
* it does not record the vat source code evaluated after the worker is started
* it does not record the incarnation number anywhere
  * although you could deduce it by counting calls to `initialize-worker`

For these, the new `initialize-worker` event should include bundleIDs for all three. Note that the helper bundles are XS-specific, but all workers will have a vat source bundleID. We should divide the data appropriately, maybe matching `options.workerOptions`. It might be appropriate to just include `workerOptions` in the event directly, then add a vat-source bundleID.

I think the transcript (as recorded in swing-store) should only have bundleIDs, *not* the full bundles. However it might make sense for a tool like `extract-transcript-from-kerneldb.js` to pull any bundles the transcript needs from `bundleStore` and write them into the generated `.sst` file, making it a stand-alone transcript artifact that no longer needs a kernelDB to run.

mhofman commented:
> I think the transcript (as recorded in swing-store) should only have bundleIDs, _not_ the full bundles. However it might make sense for a tool like `extract-transcript-from-kerneldb.js` to pull any bundles the transcript needs from `bundleStore` and write them into the generated `.sst` file, making it a stand-alone transcript artifact that no longer needs a kernelDB to run.

Agreed, and this is roughly what https://github.com/Agoric/agoric-sdk/pull/7434 does: introduce a new tool that extracts the bundles, and keep referencing bundleId in the extracted transcript file.

warner commented:
@mhofman : best reviewed commit-by-commit. The commit comments have explanations of the intended changes.

warner commented:
cc @FUDCo

mhofman commented on packages/SwingSet/test/test-controller.js:
```diff
@@ -35,6 +35,7 @@ test('load empty', async t => {
 
 async function simpleCall(t) {
   const config = {
+    defaultManagerType: 'local',
```
Nit: Maybe define a `const managerType = 'local'` and use both here and in `stateFor` ?

warner replied:
maybe, but if you were to edit that line and set it to `xsnap`, then the `workerOptions` in `stateFor()` would need to include a couple of lockdown/supervisor bundleIDs too, violating the expectation that the quick-to-modify-one-line setting would be implying

mhofman reviewed (approved):
Just a couple minor nits. This is awesome!

warner commented on packages/SwingSet/src/kernel/state/vatKeeper.js:
```diff
@@ -512,34 +527,55 @@ export function makeVatKeeper(
     } = info;
+
+    // push a save-snapshot transcript entry
+    addToTranscript({
+      d: /** @type {TDSaveSnapshot} */ ['save-snapshot'],
+      sc: [],
+      r: /** @type {TDSaveSnapshotResults} */ { status: 'ok', snapshotID },
+    });
```
huh, you're right, it works without that, must have been just me being paranoid

warner commented on packages/SwingSet/src/kernel/vat-warehouse.js:
```diff
@@ -298,6 +330,22 @@ export function makeVatWarehouse({
+      const initOpts = { source: {}, workerOptions: options.workerOptions };
+      // if the vat is somehow using a full bundle, we don't want that
+      // in the transcript: we only record bundleIDs
+      initOpts.source.bundleID = source.bundleID;
```
yeah, that would feel like multiple places each computing their own ID, which smells wrong. Also I don't want to thread a SHA256 hasher into vat-warehouse :)

mhofman replied:
Oh I was just thinking about unwrapping a `source.bundle` to get at the moduleFormat and `endoZipBase64Sha512` if any, bot to recompute a hash ;)

warner commented:
The deployment test failure revealed a bug: swingstore imports insist upon `snapshotInfo.endPos === transcriptInfo.startPos`, but now the `save-snapshot` event appears in the transcript, so the current span's `startPos` will be `snapshotInfo.endPos + 1`.

warner commented:
@FUDCo could you look at the last commit in particular? In the new approach, each old span looks like e.g.:
* 4: load-snapshot
* 5: message
* 6: message
* 7: save-snapshot

followed by a current span like:
* 8: load-snapshot
* 9: message



Which means the latest snapshot will be from `.endPos = 7`, but the latest span will have `.startPos = 8`. Previously, these two values were expected to be the same, because there was no `save-snapshot` event in the transcript (nor did it consume a deliveryNum, which is now better described as a transcriptNum).

This tripped a consistency check in the swingstore importer, but the only test that actually triggered it was the `deployment-test` (the one that runs at the very very end of a CI run, iff we're on the merge train). The swing-store `test-exportImport.js` has code which pretends to be the kernel (doing snapshot writes and span rollovers), but I missed that. If I'd properly modified that pretend-kernel code to increment the position counter in the same way the kernel now does, then `test-exportImport.js` would have caught this problem.

To reduce the chances I miss this again, I added a more direct test to swingset itself, which creates a state-sync export from a short-lived kernel, and then tries to load it back in. I tested that the new test fails if I revert the swingstore-importer consistency check to it's previous "positions must be equal" form.

Please double-check that I updated the test properly, and that you agree with my new consistency check.

FUDCo commented:
> Which means the latest snapshot will be from `.endPos = 7`, but the latest span will have `.startPos = 8`. Previously, these two values were expected to be the same, because there was no `save-snapshot` event in the transcript (nor did it consume a deliveryNum, which is now better described as a transcriptNum).

I haven't yet looked at the code (next in my queue) but I think this is a mistake.  My take is there shouldn't be anything like that happening.  The special entries you are adding to the transcript shouldn't be special from the transcriptStore's point of view; they're just entries.  Remember that endPos is exclusive and startPos is inclusive, so the endPos of one (finished) span and the start pos of the subsequent span should *always* be equal.

FUDCo reviewed (changes requested):
I'll add comments once I've read the code, but since @mhofman approved I'm preemptively clicking the "Request changes" button so this doesn't land until we sort out the startPos/endPos thing, which from your description is just broken.

warner commented:
take a look.. the check is comparing the snapshot endPos (which used to be the number just past the snapshot was taken, because it was the not-yet-used number at that moment, and is now the number *of* the snapshot, save-snapshot is now a fully-fledged member of the numberspace) against the current-transcript `startPos`. By asserting that they were equal, we were saying that there should be no deliveries after the snapshot was taken, which used to be true. But now, the snapshot should be the *last* "delivery" of the old transcript, which means the current transcript should start at one *higher* than the snapshot, hence the new `+1` check.

FUDCo commented on packages/SwingSet/src/controller/initializeSwingset.js:
```diff
@@ -327,7 +327,7 @@ export async function initializeSwingset(
   // Override the worker type if specified by the caller, to avoid having to
   // edit the config just to run everything under a different manager.
-  const defaultManagerType = SWINGSET_WORKER_TYPE || config.defaultManagerType;
+  const defaultManagerType = config.defaultManagerType || SWINGSET_WORKER_TYPE;
```
I'm not convinced this priority change is the right thing (though I'm persuaded that the use case motivating it is a real one that needs to be addressed somehow), because I think it means that some stuff that now goes through CI twice to cover both local and xsnap flavors now won't because the test's config file would now override the environment-derived controls that the CI script employs, and that consequently we'll lose some important test coverage.  I might be able to be convinced otherwise, but you'll have to sell me.

warner replied:
Good question. Here's the list of non-SwingSet test files in our tree which mention `defaultManagerType: 'xs-worker'`:

```
% rgjs defaultManagerType|grep test|grep -v SwingSet/test |grep xs-worker
ERTP/test/swingsetTests/basicFunctionality/test-basicFunctionality.js:  config.defaultManagerType = 'xs-worker';
ERTP/test/swingsetTests/splitPayments/test-splitPayments.js:  config.defaultManagerType = 'xs-worker';
governance/test/swingsetTests/committeeBinary/test-committee.js:  config.defaultManagerType = 'xs-worker';
governance/test/swingsetTests/contractGovernor/test-governor.js:  config.defaultManagerType = 'xs-worker';
notifier/test/test-publish-kit.js:    defaultManagerType: 'xs-worker',
sharing-service/test/swingsetTests/sharingService/test-sharing.js:  config.defaultManagerType = 'xs-worker';
spawner/test/swingsetTests/contractHost/test-contractHost.js:  config.defaultManagerType = 'xs-worker';
vats/test/upgrading/test-upgrade-contracts.js:    defaultManagerType: 'local', // Overridden in CI with SWINGSET_WORKER_TYPE=xs-worker
vats/test/upgrading/test-upgrade-vats.js:    defaultManagerType: 'local', // Overridden in CI with SWINGSET_WORKER_TYPE=xs-worker
zoe/test/swingsetTests/brokenContracts/test-crashingContract.js:  config.defaultManagerType = 'xs-worker';
zoe/test/swingsetTests/makeKind/test-makeKind.js:  config.defaultManagerType = 'xs-worker';
zoe/test/swingsetTests/offerArgs/test-offerArgs.js:  config.defaultManagerType = 'xs-worker';
zoe/test/swingsetTests/privateArgs/test-privateArgs.js:  config.defaultManagerType = 'xs-worker';
zoe/test/swingsetTests/runMint/test-runMint.js:  config.defaultManagerType = 'xs-worker';
zoe/test/swingsetTests/upgradeCoveredCall/test-coveredCall-service-upgrade.js:    defaultManagerType: 'xs-worker',
zoe/test/swingsetTests/zoe/test-zoe-upgrade.js:    defaultManagerType: 'xs-worker',
zoe/test/swingsetTests/zoe/test-zoe.js:  config.defaultManagerType = 'xs-worker';
```

and here's the list that mention `local`:
```
% rgjs defaultManagerType|grep test|grep -v SwingSet/test |grep local
ERTP/test/swingsetTests/ertpService/test-ertp-service-upgrade.js:    defaultManagerType: 'local',
inter-protocol/test/swingsetTests/fluxAggregator/test-fluxAggregator-service-upgrade.js:    defaultManagerType: 'local',
smart-wallet/test/swingsetTests/upgradeWalletFactory/test-walletFactory-service-upgrade.js:    defaultManagerType: 'local',
vats/test/bootstrapTests/supports.js:  config.defaultManagerType = 'local';
vats/test/upgrading/test-upgrade-contracts.js:    defaultManagerType: 'local', // Overridden in CI with SWINGSET_WORKER_TYPE=xs-worker
vats/test/upgrading/test-upgrade-vats.js:    defaultManagerType: 'local', // Overridden in CI with SWINGSET_WORKER_TYPE=xs-worker
```


The two "Overridden" comments (`vats/test/upgrading/test-upgrade-{contracts,vats}.js`) suggest that these tests probably want to be exercised both ways, and yeah, my change will reduce their test coverage.

Let's see, tests probably fall into one of the following categories:

* don't care
* want to be exercised by both `local` and `xs-worker`
* only work under `xs-worker` (they need metering, or snapshots)
* only work under `local` (they use `setup()` or some funky hybrid "steal the `syscall` object" thing)
  * these are probably only a single vat, so `vats.NAME.creationOptions.managerType` is sufficient, they don't need to use an all-vat `defaultManagerType`

We run most (all?) unit tests twice, once with the `SWINGSET_WORKER_TYPE=xs-worker` override, and once without any override. The last-priority default is `local` (because `initializeSwingset.js` has a `case undefined: config.defaultManagerType: 'local'`), so any test which omits `defaultManagerType` will get run in both ways.

So maybe what I should do is:

* try to figure out why each of the tests above expressed a preference
* if there doesn't seem to be a good reason, remove their `defaultManagerType`, so they'll get run both ways
* if the test clearly needs xsnap, retain the xs-worker, and accept that this test will be run twice in the same `xsnap` way (wasting CI time)
* it seems unlikely that any non-SwingSet tests will need `local`, but if so, retain their `defaultManagerType: 'local'`, and accept the same wasted CI time (but running it twice as `local`)

warner replied:
moved to https://github.com/Agoric/agoric-sdk/pull/7527

FUDCo commented on packages/SwingSet/src/kernel/kernel.js:
```diff
@@ -702,7 +702,7 @@ export default function buildKernel(
       const info = makeError(`${err}`);
       const results = {
-        didDelivery: true, // ok, it failed, but we did spend the time
+        didDelivery: vatID, // ok, it failed, but we did spend the time
```
Like they always say, don't do the crime if you can't do the time.

FUDCo commented on packages/SwingSet/src/kernel/state/vatKeeper.js:
```diff
@@ -463,16 +465,28 @@ export function makeVatKeeper(
+   * Generator function to return the vat's current-span transcript,
+   * one entry at a time.
    *
-   * @param {number} [startPos]  Optional position to begin reading from
```
Is it really good to lose the start position here?  Normally you only replay the current transcript span, but we at least anticipate scenarios where we need to go further back in time.

FUDCo commented on packages/SwingSet/src/kernel/state/vatKeeper.js:
```diff
@@ -506,34 +520,55 @@ export function makeVatKeeper(
     } = info;
+
+    // push a save-snapshot transcript entry
+    addToTranscript({
+      d: /** @type {TDSaveSnapshot} */ ['save-snapshot'],
+      sc: [],
+      r: { status: 'ok', snapshotID },
+    });
```
Wondering if there should be functions to construct the various special transcript entries, rather than composing the raw structures inline.

warner replied:
Maybe, but there's only one place it gets used, so there's no factoring argument in favor of a helper function.

FUDCo replied:
Yes, but there are several of these, each with the same complicated structure unfolded but with just minor variations.

warner replied:
alright, you convinced me

warner replied:
ugh, factoring them out into helper functions caused types to break

FUDCo commented on packages/SwingSet/src/kernel/vat-warehouse.js:
```diff
@@ -38,8 +42,31 @@ function recordSyscalls(origHandler) {
+function onlyRealDelivery(transcriptEntry) {
+  const dtype = transcriptEntry.d[0];
+  if (
+    dtype === 'save-snapshot' ||
+    dtype === 'load-snapshot' ||
+    dtype === 'initialize-worker' ||
+    dtype === 'shutdown-worker'
```
Might it be safer under maintenance to look for and then allow the kinds of entries that are allowed under replay rather than looking for and then rejecting the ones that are prohibited?  It's one less special `if` that we need to remember to update if we add another special entry type.

warner replied:
Hm. The list of ordinary deliveries lives in the `VatDeliveryObject` type in liveslots:

https://github.com/Agoric/agoric-sdk/blob/dde6eee8eb84b930062c962648161205520ff0dc/packages/swingset-liveslots/src/types.js#L34-L47

and currently contains 9 types: `message`, `notify`, `dropExports`, `retireExports`, `retireImports`, `changeVatOptions`, `startVat`, `stopVat`, `bringOutYourDead`. To do it right, we'd need liveslots to export a (non-type) list of normal ones. Except that different vats can be using different versions of liveslots, so if we e.g. add one in the future, then this vat-warehouse.js check would need to fetch the right list (`vatKeeper.getLiveslotsSettings`??).

It feels a bit more future-proof to have swingset reject the pseudo-delivery types that swingset has defined (since no version of liveslots will know about any of these), than to have it only accept the normal-delivery types that liveslots (which version?) defines.

But it *would* feel a bit better if swingset had a concrete list of these pseudo-delivery types, rather than vat-warehouse having a special `if` statement that is supposed to know about them. I'm not sure where that list ought to live, though, they're nominally defined in `types-internal.js`:

https://github.com/Agoric/agoric-sdk/blob/b91f3ad4c5ce51bc7c1fe6fa9a11afb1d7e335a2/packages/SwingSet/src/types-internal.js#L102-L112

but of course types aren't available for introspection by code. I suppose I could add a real export to `types-internal.js`, with a list of names? Or a `Set`, which might allow for faster lookup?

Anyways, I'm inclined to defer this to the future.

FUDCo replied:
Couldn't there simply be an `isDeliverable` predicate?

warner replied:
If `isDeliverable` would live in `vat-warehouse.js`, then that's basically what `onlyRealDelivery` is, except that it takes a transcriptEntry and returns a delivery object:
```js
/**
 * @param {TranscriptEntry} transcriptEntry
 * @returns {VatDeliveryObject}
 */
function onlyRealDelivery(transcriptEntry) {
  const dtype = transcriptEntry.d[0];
  if (
    dtype === 'save-snapshot' ||
    dtype === 'load-snapshot' ||
    dtype === 'initialize-worker' ||
    dtype === 'shutdown-worker'
  ) {
    throw Fail`replay should not see ${dtype}`;
  }
  return transcriptEntry.d;
}
```

If it should live in liveslots, then we have the question of *which* (version of) liveslots it lives in. The kernel doesn't talk directly to liveslots, it talks to a vat which may or may not contain (some version of) liveslots. So there's no simple way for `packages/liveslots` to export a predicate that the kernel can query (no liveslots code is ever evaluated inside the kernel's Compartment).

FUDCo replied:
I would expect this to be a kernel predicate, as it's about purely kernel-side stuff that the kernel wants to avoid delivering to the vat.  My concern was simply that an allow-list is safer than a deny-list if we introduce another kernel-side special pseudo-deliverable thingie in the future.  If somebody does that and misses the need to update the `onlyRealDelivery` function here, then we could end up dropping something into a vat that could break it.  The list of things that _can_ be delivered into a vat is likely to be more stable, as there's no way to introduce a new actual deliverable type without coordination with the vat anyway.

To reiterate, this is purely a hygiene thing and doesn't touch on fundamental semantics.

FUDCo commented on packages/swing-store/src/swingStore.js:
```diff
@@ -1073,7 +1073,7 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
         Fail`missing current snapshot for vat ${q(vatID)}`;
     } else {
-      snapshotInfo.endPos === transcriptInfo.startPos ||
+      snapshotInfo.endPos + 1 === transcriptInfo.startPos ||
```
On deeper thought I'm reasonably convinced this is a mistake and that the old invariant check is the correct one.  The thing that's adding the snapshot entry into the transcript should be incrementing the `endPos` value the same as any other addition to the transcript.  After looking at the code that's adding these special entries, I'm puzzled by why that's not happening, but I think something is wrong there.

mhofman replied:
I gave this some thoughts, and I believe this change is logically correct. However it highlights a potential problem with this PR
When we create a snapshot, we record the current transcript position `endPos`.
Now with this PR when the snapshot is created, we also insert a transcript entry with the snapshot details, then roll over the span.
It used to be that the new span started at that `endPos`, but now it starts at that `endPos + 1` because of the extra transcript entry.

Either we accept that difference (which has the nice property that the snapshot position is the position of the corresponding `snapshot-save` transcript entry), or we record the snapshot position as `endPos + 1` instead to keep the snapshotPos === next transcript span start.

FUDCo replied:
I just don't get this.  The `endPos` isn't part of the snapshot, so I don't see how the timing of the snapshot enters into the discussion at all.

FUDCo reviewed (approved):
At this point my concerns have been resolved to my satisfaction.
